File: sources/iOS/ios-communications/Sources/iOSCommunications/ble/api/model/BleDeviceSession.swift
import Foundation
import CoreBluetooth
public enum BleDisconnectReason: Equatable {
    case insufficientEncryption
    case encryptionTimedOut
    case pairingInformationRemoved
    case uuidNotAllowed
    case unknown
}

/// Device-control command that a pending expected disconnect is following.
public enum BleDeviceCommand: Equatable {
    case restart
    case factoryReset
    case warehouseSleep
    case hibernate
    case turnOff
}

@objc open class BleDeviceSession: NSObject {

    public enum DeviceSessionState {
        case sessionClosed, sessionOpening, sessionOpenPark, sessionOpen, sessionClosing

        public func description() -> String {
            switch self {
            case .sessionClosed:  return "sessionClosed"
            case .sessionOpening: return "sessionOpening"
            case .sessionOpenPark: return "sessionOpenPark"
            case .sessionOpen:    return "sessionOpen"
            case .sessionClosing: return "sessionClosing"
            }
        }
    }

    // Put initial value that is well below actual BLE sensitivity.
    public var rssi: Int = -120

    public enum ConnectionType {
        /// connection attempt is directly requested after disconnection
        case directConnection
        /// connection attempt is requested after first advertisement head
        case connectFromAdvertisementHead
    }

    public let address: UUID
    public let advertisementContent: BleAdvertisementContent
    public var state = DeviceSessionState.sessionClosed
    public var error: Error? = nil
    public var previousState = DeviceSessionState.sessionClosed

    /// by default connect only from adv head
    public var connectionType = ConnectionType.connectFromAdvertisementHead
    var gattClients = [BleGattClientBase]()
    public private(set) var hasEstablishedConnection = false
    public private(set) var securityRecoveryAttempts = 0
    public private(set) var securityRecoveryExhausted = false

    // Device-control command (restart, factory reset, warehouse sleep, hibernate, turn off)
    // whose success path was already reached; the disconnect that follows is expected.
    public private(set) var pendingDeviceCommand: BleDeviceCommand? = nil
    private var deviceCommandDisconnectDeadline: Date? = nil

    /// Call right after a device-control command's success path to mark the disconnect that
    /// is expected to follow as non-actionable. A still-valid pending mark is kept as-is rather
    /// than overwritten, since once accepted the device is already on its way out and a second
    /// command issued moments later is unlikely to be the real cause.
    public func markExpectedDeviceCommandDisconnect(_ command: BleDeviceCommand, validFor: TimeInterval = 60) {
        guard expectedDeviceCommandIfStillValid() == nil else { return }
        pendingDeviceCommand = command
        deviceCommandDisconnectDeadline = Date().addingTimeInterval(validFor)
    }

    public func expectedDeviceCommandIfStillValid() -> BleDeviceCommand? {
        guard let command = pendingDeviceCommand, let deadline = deviceCommandDisconnectDeadline, Date() <= deadline else {
            return nil
        }
        return command
    }

    public func clearExpectedDeviceCommandDisconnect() {
        pendingDeviceCommand = nil
        deviceCommandDisconnectDeadline = nil
    }

    public func markConnectionEstablished() {
        hasEstablishedConnection = true
        securityRecoveryAttempts = 0
        securityRecoveryExhausted = false
        clearExpectedDeviceCommandDisconnect()
    }

    @discardableResult
    public func recordSecurityRecoveryAttempt(maxAttempts: Int = 3) -> Bool {
        securityRecoveryAttempts += 1
        securityRecoveryExhausted = securityRecoveryAttempts >= maxAttempts
        return securityRecoveryExhausted
    }

    public init(_ addr: UUID, advertisementContent: BleAdvertisementContent? = nil) {
        self.advertisementContent = advertisementContent ?? BleAdvertisementContent()
        self.address = addr
    }

    /// helper to return BleGattClientBase instance based on service uuid
    public func fetchGattClient(_ serviceUuid: CBUUID) -> BleGattClientBase? {
        return gattClients.first(where: { $0.serviceBelongsToClient(serviceUuid) })
    }

    public func isConnectable() -> Bool {
        fatalError("not implemented")
    }

    public func retryServiceDiscovery() {
        fatalError("not implemented")
    }

    /// Monitor services discovered on the device.
    ///
    /// - Parameter checkConnection: check current connection
    /// - Returns: AsyncThrowingStream emitting discovered service UUIDs, completing when all services are ready, or throwing on error
    public func monitorServicesDiscovered(_ checkConnection: Bool) -> AsyncThrowingStream<CBUUID, Error> {
        fatalError("not implemented")
    }

    public var disconnectedDueRemovedPairing: Bool {
        // A forgotten bond can surface either as an explicit peer-removed
        // error or as an ATT request that can no longer be encrypted.
        switch self.error?.bleDisconnectReason {
        case .pairingInformationRemoved:
            return true
        case .insufficientEncryption:
            return hasEstablishedConnection || securityRecoveryExhausted
        default:
            return false
        }
    }
}

File: sources/Android/android-communications/library/src/main/java/com/polar/androidcommunications/api/ble/model/BleDeviceSession.kt
package com.polar.androidcommunications.api.ble.model

import android.bluetooth.BluetoothDevice
import com.polar.androidcommunications.api.ble.model.advertisement.BleAdvertisementContent
import com.polar.androidcommunications.api.ble.model.advertisement.BlePolarHrAdvertisement
import com.polar.androidcommunications.api.ble.model.gatt.BleGattBase
import kotlinx.coroutines.Deferred
import java.util.UUID
import java.util.concurrent.TimeUnit

/**
 * Bluetooth le device class, contains all essential api's for sufficient usage of bluetooth le device
 */
abstract class BleDeviceSession
/**
 * Methods
 *
 *
 * Class constructor
 */
protected constructor() {
    enum class DisconnectReason {
        NONE,
        CONNECTION_LOST,
        INSUFFICIENT_ENCRYPTION,
        INSUFFICIENT_AUTHENTICATION,
        // Reserved for parity with iOS's CBError.encryptionTimedOut; no known Android GATT
        // status currently maps to this, so it is never produced by GattCallback today. If a
        // producer is added, mirror iOS: only RETRY_CONNECTION when the session never reached
        // markConnectionEstablished()-equivalent state, otherwise NONE (routine reconnect).
        ENCRYPTION_TIMEOUT,
        PAIRING_INFORMATION_REMOVED,
        PAIRING_NEGOTIATION_FAILED,
        SERVICE_DISCOVERY_FAILED,
        GATT_ERROR,
        // The remote device actively terminated the link (e.g. HCI "remote user terminated" /
        // "remote power off"), as opposed to the connection merely timing out.
        PEER_TERMINATED,
        // Disconnect is the expected outcome of a device-control command (restart, factory
        // reset, warehouse sleep, hibernate, turn off) whose success path was already reached.
        DEVICE_COMMAND,
        UNKNOWN
    }

    /** Device-control command that a DEVICE_COMMAND disconnect is expected to follow. */
    enum class DeviceCommand {
        RESTART,
        FACTORY_RESET,
        WAREHOUSE_SLEEP,
        HIBERNATE,
        TURN_OFF
    }

    var disconnectReason: DisconnectReason = DisconnectReason.NONE
        protected set
    var disconnectStatus: Int? = null
        protected set
    var isIntentionalDisconnect: Boolean = false
        protected set
    // Deadline until which a pending DEVICE_COMMAND marking is honored; prevents a stale mark
    // from mislabeling a later, unrelated disconnect if the expected one never arrives.
    var deviceCommandDisconnectDeadlineMs: Long = 0L
        protected set
    var pendingDeviceCommand: DeviceCommand? = null
        protected set
    var wasBondedAtConnection: Boolean = false
        protected set
    var hasEstablishedBond: Boolean = false
        protected set
    var securityRecoveryAttempts: Int = 0
        protected set
    var securityRecoveryExhausted: Boolean = false
        protected set

    fun markBondedAtConnection(bonded: Boolean) {
        wasBondedAtConnection = bonded
        if (bonded) {
            hasEstablishedBond = true
        }
        securityRecoveryAttempts = 0
        securityRecoveryExhausted = false
    }

    fun recordSecurityRecoveryAttempt(maxAttempts: Int = 2): Boolean {
        securityRecoveryAttempts += 1
        securityRecoveryExhausted = securityRecoveryAttempts >= maxAttempts
        return securityRecoveryExhausted
    }

    fun markDisconnect(reason: DisconnectReason, status: Int? = null) {
        disconnectReason = reason
        disconnectStatus = status
    }

    fun markIntentionalDisconnect() {
        isIntentionalDisconnect = true
        markDisconnect(DisconnectReason.CONNECTION_LOST)
    }

    /** Call right after a device-control command's success path (e.g. GATT write acked) to mark the disconnect that is expected to follow as non-actionable. */
    fun markExpectedDeviceCommandDisconnect(command: DeviceCommand, validForMs: Long = 60_000L) {
        // Once accepted, a reset-family command already has the device on its way out; a second
        // such command issued before that disconnect arrives is unlikely to be the real cause, so
        // keep attributing the upcoming disconnect to whichever command was marked first.
        if (isExpectedDeviceCommandDisconnectStillValid()) {
            return
        }
        pendingDeviceCommand = command
        deviceCommandDisconnectDeadlineMs = System.currentTimeMillis() + validForMs
        markDisconnect(DisconnectReason.DEVICE_COMMAND)
    }

    fun isExpectedDeviceCommandDisconnectStillValid(): Boolean =
        disconnectReason == DisconnectReason.DEVICE_COMMAND && System.currentTimeMillis() <= deviceCommandDisconnectDeadlineMs

    fun clearDisconnectReason() {
        disconnectReason = DisconnectReason.NONE
        disconnectStatus = null
        isIntentionalDisconnect = false
        deviceCommandDisconnectDeadlineMs = 0L
        pendingDeviceCommand = null
    }

    /**
     * Connection state
     */
    enum class DeviceSessionState {
        /**
         * Disconnected state
         */
        SESSION_CLOSED,

        /**
         * Connection attempting/connecting at the moment
         */
        SESSION_OPENING,

        /**
         * Device is disconnected, but is waiting for advertisement head or ble power on for reconnection
         */
        SESSION_OPEN_PARK,

        /**
         * Device is connected
         */
        SESSION_OPEN,

        /**
         * Disconnecting at the moment
         */
        SESSION_CLOSING,
    }

    /**
     * @return get current state (DeviceSessionState.SESSION_XXXXX)
     */
    /**
     * Members
     */
    var sessionState: DeviceSessionState = DeviceSessionState.SESSION_CLOSED

    /**
     * @return state before current one
     */
    var previousState: DeviceSessionState = DeviceSessionState.SESSION_CLOSED
        protected set
    protected lateinit var clients: Set<BleGattBase>

    // needs to be set by 'upper' class
    var advertisementContent: BleAdvertisementContent = BleAdvertisementContent()

    /**
     * @return advertisement content object
     */
    val getAdvertisementContent: BleAdvertisementContent
        get() = advertisementContent

    /**
     * Set advertisement content object
     */
    /*fun setAdvertisementContent(content: BleAdvertisementContent) {
        advertisementContent = content
    }*/

    /**
     * @return current connection uuid's needed for connection attempt
     */
    var connectionUuids: MutableList<String> = ArrayList()
        /**
         * @param uuids set connection uuid's
         */
        set(uuids) {
            field.clear()
            field.addAll(uuids)
        }

    /**
     * @param time     max time for the last advertisement event
     * @param timeUnit desired timeunit
     * @return true, device is still somewhat considered "alive"
     */
    fun isDeviceAlive(time: Long, timeUnit: TimeUnit): Boolean {
        val deltaTime =
            (System.currentTimeMillis() / 1000L) - advertisementContent.advertisementTimeStamp
        val sessionState = sessionState
        return (sessionState == DeviceSessionState.SESSION_OPEN || sessionState == DeviceSessionState.SESSION_OPENING || sessionState == DeviceSessionState.SESSION_CLOSING) ||
                (advertisementContent.getAdvertisementData().size != 0 && deltaTime <= timeUnit.toSeconds(
                    time
                ))
    }

    /**
     * @param time     max time for the last advertisement event
     * @param timeUnit desired timeunit
     * @return true if device is advertising in the required timespan
     */
    fun isAdvertising(time: Long, timeUnit: TimeUnit): Boolean {
        val deltaTime =
            (System.currentTimeMillis() / 1000L) - advertisementContent.advertisementTimeStamp
        return advertisementContent.getAdvertisementData().isNotEmpty() && deltaTime <= timeUnit.toSeconds(
            time
        )
    }

    /**
     * @return true if device is in non-connectable advertisement<BR></BR>
     * Notes: Depending on Android API version,it's impossible to know this.<BR></BR>
    * So this would work only for Polar Devices that follows Polar BLE Advertising data spec
     */
    abstract val isNonConnectableAdvertisement: Boolean

    val isConnectableAdvertisement: Boolean
        /**
         * @return true if device is connectable advertisement<BR></BR>
         * Notes: Depending on Android API version,it's impossible to know this.<BR></BR>
         * So this would work only for Polar Devices that follows Polar BLE Advertising data spec
         */
        get() = !isNonConnectableAdvertisement

    /**
     * @return bluetooth device address in string format
     */
    abstract val address: String?

    /**
     * start pairing and optionally bonding procedure with the device
     * @return Either produces complete for already bonded device or starts pairing/bonding procedure
     * or onError on failure case
     */
    abstract suspend fun authenticate()

    /**
     * @return true if session is bonded device
     */
    abstract val isAuthenticated: Boolean

    /**
     * Monitor services discovered
     *
     * @return Deferred result
     */
    abstract fun monitorServicesDiscovered(checkConnection: Boolean): Deferred<List<UUID>>

    /**
     * @return true if current gatt cache clear was ok
     */
    abstract fun clearGattCache(): Boolean

    /**
     * @return Deferred result for rssi values
     */
    abstract fun readRssiValue(): Deferred<Int>?

    /**
     * Checks if session indicates pairing problem.
     * This is a helper method to check if the session is in a state where pairing problems are likely, for example due to failed authentication attempts
     * or other issues that may arise during the pairing process.
     * @return Boolean indicating if the session indicates a pairing problem.
     * True if there are indications of a pairing problem, false otherwise. Int value contains Bluetooth error code if there is a pairing problem, otherwise it is -1 (not set).
     */
    abstract fun getIndicatesPairingProblem():  Pair<Boolean, Int>

    /**
     * @param uuid service uuid
     * @return get a specific client from the list
     */
    fun fetchClient(uuid: UUID): BleGattBase? {
        for (serviceBase in clients) {
            if (serviceBase.serviceBelongsToClient(uuid)) {
                return serviceBase
            }
        }
        return null
    }

    /**
     * Helper to combine all available/desired clients ready.
     *
     * @param checkConnection check initial connection
     * @throws Throwable if any client fails to become ready
     */
    suspend fun clientsReady(checkConnection: Boolean) {
        val uuids = monitorServicesDiscovered(checkConnection).await()
        for (uuid in uuids) {
            fetchClient(uuid)?.clientReady(checkConnection)
        }
    }

    /**
     * @return android bluetooth device instance
     */
    abstract val bluetoothDevice: BluetoothDevice?

    val name: String
        // adv data getter helpers
        get() = advertisementContent.name

    val polarDeviceId: String
        /**
         * @return polar device id
         */
        get() = advertisementContent.polarDeviceId

    open val polarDeviceType: String
        /**
         * @return polar device type
         */
        get() = advertisementContent.polarDeviceType

    val polarDeviceIdInt: Long
        /**
         * @return polar device id in int
         */
        get() = advertisementContent.polarDeviceIdInt

    val medianRssi: Int
        /**
         * @return current median rssi value
         */
        get() = advertisementContent.medianRssi

    val rssi: Int
        /**
         * @return instant rssi value
         */
        get() = advertisementContent.rssi

    val blePolarHrAdvertisement: BlePolarHrAdvertisement
        /**
         * @return polar hr advertisement content @see BlePolarHrAdvertisement
         */
        get() = advertisementContent.polarHrAdvertisement
}
